engine/runtime/src/main/java/org/enso/compiler/pass/analyse/ExportSymbolAnalysis.java

```diff
@@ -0,0 +1,138 @@
+import scala.jdk.javaapi.CollectionConverters;
+
+/**
+ * This pass ensures that all the symbols that are exported exist. If not, an IR error is generated.
+ */
+public class ExportSymbolAnalysis implements IRPass {
+  public static final ExportSymbolAnalysis INSTANCE = new ExportSymbolAnalysis();
+  private UUID uuid;
+
+  private ExportSymbolAnalysis() {}
+
+  @Override
+  public UUID key() {
+    return null;
+  }
+
+  @Override
+  public void org$enso$compiler$pass$IRPass$_setter_$key_$eq(UUID v) {
+    this.uuid = v;
+  }
+
+  @Override
+  public Seq<IRPass> precursorPasses() {
+    List<IRPass> passes = List.of(
+        BindingAnalysis$.MODULE$,
+        ImportSymbolAnalysis$.MODULE$
+    );
+    return CollectionConverters.asScala(passes).toList();
+  }
+
+  @Override
+  public Seq<IRPass> invalidatedPasses() {
+    return ScalaConversions.nil();
+  }
+
+  @Override
+  public Module runModule(Module moduleIr, ModuleContext moduleContext) {
+    List<Export> exportErrors = new ArrayList<>();
+    var bindingsMap = (BindingsMap) moduleIr.passData().get(BindingAnalysis$.MODULE$).get();
+
+    moduleIr.exports().foreach(export -> switch (export) {
+      case Export.Module exportMod -> {
+        var exportNameParts = exportMod.name().parts();
+        var symbolName = exportMod.name().parts().last();
+        assert exportNameParts.size() > 1;
+        var moduleOrTypeName = exportNameParts.apply(exportNameParts.size() - 2);
+        var foundResolvedExp = findResolvedExportForIr(export, bindingsMap);
+        if (foundResolvedExp == null) {
+          exportErrors.add(
+              ImportExport.apply(
+                  symbolName,
+                  new ImportExport.SymbolDoesNotExist(symbolName.name(), moduleOrTypeName.name()),
+                  ImportExport.apply$default$3(),
+                  ImportExport.apply$default$4()
+              )
+          );
+        } else {
+          if (exportMod.onlyNames().isDefined()) {
+            assert exportMod.onlyNames().isDefined();
+            var exportedSymbols = exportMod.onlyNames().get();
+            exportedSymbols.foreach(exportedSymbol -> {
+              var foundSymbols = foundResolvedExp.target().findExportedSymbolsFor(exportedSymbol.name());
+              if (foundSymbols.isEmpty()) {
+                exportErrors.add(
+                    ImportExport.apply(
+                        exportedSymbol,
+                        new ImportExport.SymbolDoesNotExist(exportedSymbol.name(), moduleOrTypeName.name()),
+                        ImportExport.apply$default$3(),
+                        ImportExport.apply$default$4()
+                    )
+                );
+              }
+              return null;
+            });
+          }
+        }
+        yield null;
+      }
+      default -> export;
+    });
+    return moduleIr.copy(
+        moduleIr.imports(),
+        exportErrors.isEmpty() ? moduleIr.exports() : CollectionConverters.asScala(exportErrors).toList(),
```

Comment:
If there are no `exportErrors`, do no copy, return original `moduleIr`.

Reply:
Good point. Refactored in 0b8b8872deeb01de6814abd7d465ab8a73fac7f4